lib/pure/ospaths.nim

```diff
@@ -157,64 +158,173 @@ const
     ## The character which separates the base filename from the extension;
     ## for example, the '.' in ``os.nim``.
 
+# TODO: MOVE to strutils, sequtils, or algorithms
+proc countWhile*(str: string, pred : proc(a:char):bool, prefix = true): int =
+  ## returns the number of elements in ``str`` that satisfy ``pred`` predicate.
+  ## without interruption, starting from beginning (when ``prefix`` is true)
+  ## or end (when ``prefix`` is false)
+  runnableExamples:
+    import sugar
+    doAssert countWhile("abc", a=>a == '/') == 0
+    doAssert countWhile("//abc", a=>a == '/') == 2
+    doAssert countWhile("//", a=>a == '/') == 2
+    doAssert countWhile("", a=>a == '/') == 0
+
+    doAssert countWhile("abc", a=>a == '/', prefix = false) == 0
+    doAssert countWhile("abc//", a=>a == '/', prefix = false) == 2
+    doAssert countWhile("//", a=>a == '/', prefix = false) == 2
+    doAssert countWhile("", a=>a == '/', prefix = false) == 0
+  let lenS = len(str)
+  if prefix:
+    for i in 0..<lenS:
+      if not pred(str[i]): return i
+  else:
+    for i in 0..<lenS:
+      if not pred(str[lenS - i - 1]): return i
+  return lenS
+
+# TODO: once slices are supported, define instead `rootPrefix`
+proc rootPrefixLength*(path: string) : int =
+  ## Returns the length of the prefix that makes ``path`` absolute, or 0
+  ## if ``path`` is relative.
+  runnableExamples:
+    doAssert "foo".rootPrefixLength == 0
+    doAssert "/foo".rootPrefixLength == 1
+    doAssert "//foo".rootPrefixLength == 2
+    when doslikeFileSystem:
+      doAssert r"\foo".rootPrefixLength == 1
+      doAssert r"/foo".rootPrefixLength == 1
+      doAssert r"C:".rootPrefixLength == 2
+      doAssert r"C:\foo".rootPrefixLength == 3
+      doAssert r"//foo".rootPrefixLength == 2
+      doAssert r"C:\\foo".rootPrefixLength == 4
```

Comment:
As far as I'm concerned this is abusing ``runnableExamples`` as exhaustive testing again. Examples should be one or two (one per platform). Not a long list of edge cases.

Reply:
fixed; moved out from runnableExamples

Reply:
fixed; moved out of runnableExamples